game/src/main/kotlin/jake2/game/Door.kt

```diff
@@ -0,0 +1,726 @@
+    VectorCopy(self.s.origin, self.moveinfo.end_origin)
+    VectorCopy(self.pos2, self.moveinfo.end_angles)
+
+    if (self.spawnflags and 16 != 0)
+        self.s.effects = self.s.effects or Defines.EF_ANIM_ALL
+
+    // to simplify logic elsewhere, make non-teamed doors into a team of one
+    if (self.team == null)
+        self.teammaster = self
+
+    game.gameImports.linkentity(self)
+
+    self.think.nextTime = game.level.time + Defines.FRAMETIME
+    if (self.health != 0 || self.targetname != null)
+        self.think.action = Think_CalcMoveSpeed
+    else
+        self.think.action = spawnTouchTrigger
+    true
+
+}
+
+private val doorBlocked = registerBlocked("door_blocked") { self, obstacle, game ->
+    // if not a monster or player
+    if ((0 == obstacle.svflags and Defines.SVF_MONSTER) && obstacle.client == null) {
+        // give it a chance to go away on its own terms (like gibs)
+        GameCombat.T_Damage(
+            obstacle, self, self, Globals.vec3_origin,
+            obstacle.s.origin, Globals.vec3_origin, 100000, 1, 0,
+            GameDefines.MOD_CRUSH, game
+        )
+        // if it's still there, nuke it
+        // fixme: was != null before
+        if (obstacle.inuse)
+            GameMisc.BecomeExplosion1(obstacle, game)
+        return@registerBlocked
+    }
+
+    GameCombat.T_Damage(
+        obstacle, self, self, Globals.vec3_origin,
+        obstacle.s.origin, Globals.vec3_origin, self.dmg, 1, 0,
+        GameDefines.MOD_CRUSH, game
+    )
+
+    if (self.spawnflags and DOOR_CRUSHER != 0)
+        return@registerBlocked
+
+    // if a door has a negative wait, it would never come back if
+    // blocked,
+    // so let it just squash the object to death real fast
+
+    if (self.moveinfo.wait >= 0) {
+        var team: SubgameEntity?
+        if (self.moveinfo.state == GameFunc.STATE_DOWN) {
+            team = self.teammaster
+            while (team != null) {
+                doorOpening(team, team.activator, game)
+                team = team.teamchain
+            }
+        } else {
+            team = self.teammaster
+            while (team != null) {
+                doorClosing.think(team, game)
+                team = team.teamchain
+            }
+        }
+    }
+
+}
+
+val doorOpenUse = registerUse("door_use") { self, other, activator, game ->
+
+    if (self.flags and GameDefines.FL_TEAMSLAVE != 0)
+        return@registerUse
+
+    if (self.spawnflags and GameFunc.DOOR_TOGGLE != 0) {
+        if (self.moveinfo.state == GameFunc.STATE_UP || self.moveinfo.state == GameFunc.STATE_TOP) {
+            // trigger all paired doors
+            var team: SubgameEntity? = self
+            while (team != null) {
+                team.message = null
+                team.touch = null
+                doorClosing.think(team, game)
+                team = team.teamchain
+            }
+            return@registerUse
+        }
+    }
+
+    // trigger all paired doors
```

Comment:
Could it be extracted to a method? And maybe its name could help to understand what is happening